### PR TITLE
Support CUDA Graph for MoE models

### PR DESCRIPTION
**Pull request by buptzyb**
# Description

Different from non-MoE models like llama2, MoE models have dynamic shaped activations in FFN layers, so one cudagraph can only capture a part of one transformer layer, instead of covering the whole layer. We call this a "breaking-layer" cudagraph mode. This PR adds breaking-layer cudagraph supports for MoE models on the TE side, and fixes several related bugs in TE.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Add `is_initialized()` method in CudaRNGStatesTracker to align with what is already done in MCore.
- Fix wrong `per_callable_module_params` order bug in `_make_graphed_callables` when `_order` is given.
- Fix warmup argument mismatch bug in `_make_graphed_callables` when `_order` is given.
- Fix fp8 accuracy issue by adding `fp8_group` argument to `make_graphed_callables()` and modifing `is_first_microbatch`, `skip_fp8_weight_update` and `fp8_meta` code.
- Support MoE models cudagraph by filtering graphed TE modules and model weights during warmup.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/graph.py [hunk 1]
```diff
@@ -200,21 +204,96 @@ def _make_graphed_callables(
...
+
+    def get_no_grads(static_input_surface, grad_inputs, func_idx):
+        grad_index = 0
+        none_grads = []
+        for i in range(len(static_input_surface)):
+            if static_input_surface[i].requires_grad:
+                if grad_inputs[grad_index] is None and i >= len(flatten_sample_args[func_idx]):
+                    none_grads.append(i)
+                grad_index += 1
+        return set(none_grads)
```
Comment (timmoon10) [inline transformer_engine/pytorch/graph.py:256]: If I understand correctly, you are checking for params that recieve no grads and removing them from the static input surface. Is this safe? A param might still be required in the forward compute even if it is disconnected from the autograd graph, e.g. by calling `tensor.detach()` or using `torch.no_grad()`.

File: transformer_engine/pytorch/graph.py [hunk 2]
```diff
@@ -459,6 +538,17 @@ def new_fwd(*user_args, **user_kwargs):
...
+                                    # Only Set the FP8 meta for the modules included by forward
+                                    continue
+                                fp8_recipe = FP8GlobalStateManager.get_fp8_recipe()
+                                if (
+                                    not fp8_recipe.fp8_mha
+                                    and not fp8_recipe.fp8_dpa
+                                    and hasattr(m, "attention_dropout")
+                                    and m.deterministic
+                                ):
+                                    continue
```
Comment (timmoon10) [inline transformer_engine/pytorch/graph.py:551]: Why are we skipping the FP8 scale update logic for this case?
```suggestion
```

**Review by timmoon10 — COMMENTED** [review]
Overall I don't see glaring issues, although I have questions and stylistic suggestions.

@ptrendx @ksivaman @sbhavani What is our priority for this feature? The custom Mcore logic in `make_graphed_callables` is already messy and fragile, and this PR does exacerbate those problems. We should explicitly make a decision before committing to maintain it.

File: transformer_engine/pytorch/graph.py (hunk 2 above)
Comment (yifeis-nv) [reply transformer_engine/pytorch/graph.py:551]: This is for a deterministic test with FP8. Even without CUDA graphs, we find that FP8_DPA leads to random output, even when set to deterministic mode. With CUDA graphs, we not only need to set DPA to BF16, but also skip the FP8_meta update for DPA. After that, we will finally obtain a stable output under the deterministic test.

File: transformer_engine/pytorch/module/layernorm_linear.py [hunk 3]
```diff
@@ -1206,7 +1209,7 @@ def forward(
                 self.apply_bias and not self.gemm_bias_unfused_add,
                 self.eps,
                 is_first_microbatch,
-                self.fp8,
+                False if self.fp8 is None else self.fp8,
```
[the remark this answers is not in the document]
Comment (buptzyb) [reply transformer_engine/pytorch/module/layernorm_linear.py]: Thanks, removed.

File: transformer_engine/pytorch/graph.py (hunk 1 above)
Comment (buptzyb) [reply transformer_engine/pytorch/graph.py]: Make sense, thanks for pointing out.

TE always adds all parameters in callable as cudagraph inputs, so here I want to filter out the parameters that are not really covered by cudagraph.

Now I changed to using hooks to record the parameters that are actually accessed in warmup, together with the TE modules recorder hook.

File: transformer_engine/pytorch/graph.py [hunk 4]
```diff
@@ -200,21 +204,96 @@ def _make_graphed_callables(
...
                     grad_outputs=tuple(torch.empty_like(o) for o in outputs if o.requires_grad),
                     only_inputs=True,
                     allow_unused=allow_unused_input,
                 )
+                if no_grad_weights is None:
+                    no_grad_weights = get_no_grads(static_input_surface, grad_inputs, func_idx)
                 del outputs, grad_inputs
+            for module in func.modules():
+                if hasattr(module, "is_first_microbatch"):
+                    module.is_first_microbatch = True
```
[the remark this answers is not in the document]
Comment (yifeis-nv) [reply transformer_engine/pytorch/graph.py:268]: Thank you for the reminder. I also believe that modifications are needed here. The change we originally intended to address is the issue where the fp8 weight caching behavior in MoE leads to different behaviors for the first microbatch compared to other microbatches. If we do not include this piece of code, the warmup process will update `is_first_microbatch` to False, causing all captured graphs to exhibit non-first microbatch behavior, which does not align with our requirements. Therefore, we chose to reset this parameter after warmup.
In summary, our requirement is either to prevent `is_first_microbatch` from being updated during warmup or to reset `is_first_microbatch` after warmup. Choosing the former may require adding a flag to inform MoE that it is currently in the warmup phase, while choosing the latter might necessitate making this code a bit more general. Do you have any input on the modification plan that could serve as a reference for us?

**Comment by buptzyb** [comment]
> Have you tested that it works with Mcore?
Yes, we also made some changes in Mcore to enable MoE cudagraph. You can refer to issue 193 in our Megatron-LM repo.

File: transformer_engine/pytorch/graph.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/graph.py:268]: I see, so this is Mcore-specific logic. It's uncomfortable that it's made its way into TE, but it's a tricky problem and I can't think of a better solution either, at least without significant changes in Mcore.

We should document what this is doing, especially for TE developers with no knowledge of Mcore.

File: transformer_engine/pytorch/graph.py (hunk 2 above)
Comment (timmoon10) [reply transformer_engine/pytorch/graph.py:533]: It seems like we're covering up a correctness bug. This may also affect convergence since we are no longer doing amax reductions within MHA (e.g. for the FP8 cast before `qkv` and `proj`).

File: transformer_engine/pytorch/graph.py (hunk 1 above)
Comment (timmoon10) [reply transformer_engine/pytorch/graph.py]: The new approach is much cleaner and easier to understand.

File: transformer_engine/pytorch/graph.py (hunk 2 above)
Comment (yifeis-nv) [reply transformer_engine/pytorch/graph.py:533]: It makes sense, and I think we should '`continue`' without fp8_mha and fp8_dpa, regardless of whether the test is deterministic or not.

Comment (timmoon10) [reply transformer_engine/pytorch/graph.py:533]: Actually, this change will cause correctness issues. We do a max all-reduce on the amaxes so that FP8 scaling factors are synchronized over the TP group. If we skip the amax reduction, then the MHA's TP communication will be wrong. In particular, we'll all-gather FP8 inputs to the `qkv` GEMM, but the scaling factors will be different for each TP rank.

Comment (timmoon10) [reply transformer_engine/pytorch/graph.py:533]: As far as I can tell, this logic is only relevant to `DotProductAttention`. We could make intent much more obvious with:
```suggestion
                                if (
                                    isinstance(m, DotProductAttention)
                                    and not fp8_recipe.fp8_mha
                                    and not fp8_recipe.fp8_dpa
                                ):
                                    # Don't need to update FP8 meta for non-FP8 DPA
                                    continue
```

File: transformer_engine/pytorch/graph.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (yifeis-nv) [reply transformer_engine/pytorch/graph.py:270]: Done. I try to explain the reason for this MoE specific logic.

File: transformer_engine/pytorch/graph.py (hunk 2 above)
Comment (yifeis-nv) [reply transformer_engine/pytorch/graph.py]: Thanks! I have modified this.

**Comment by buptzyb** [comment]
Hi @timmoon10 , do you have more suggestions on this PR?

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by yaox12** [comment]
Merge this PR since pipeline 20710146 passed and Tim approved.
